### PR TITLE
Add Pytorch-TensorRT support, improved tests and bugfix

### PR DESCRIPTION
**Pull request by valeriosofi**

### CODE REVIEW

File: nebullvm/config.py [hunk 1]
```diff
@@ -43,5 +43,3 @@
     "tflite_model": "tf_model.tflite",
     "tf_model": "tf_model.h5",
 }
-
-SAVE_DIR_NAME = "optimized_model"
```
Comment (morgoth95) [inline nebullvm/config.py:47]: Why are you reverting this? We don't want to save the model files into a directory anymore? Wouldn't this lead to an issue while saving large ONNX models?

File: nebullvm/inference_learners/tensor_rt.py [hunk 2]
```diff
@@ -248,6 +247,59 @@ def load(cls, path: Union[Path, str], **kwargs):
         )
 
 
+class PytorchTensorRTInferenceLearner(PytorchBaseInferenceLearner):
+    MODEL_NAME = "model_optimized.pt"
+
+    def __init__(
+        self, torch_model: torch.jit.ScriptModule, dtype: torch.dtype, **kwargs
```
Comment (morgoth95) [inline nebullvm/inference_learners/tensor_rt.py:254]: Does TensorRT need a TorchScript module as input?

File: nebullvm/optimizers/tensor_rt.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 from pathlib import Path
 from typing import List, Tuple, Optional, Callable
+import warnings
```
Comment (morgoth95) [inline nebullvm/optimizers/tensor_rt.py:3]: sort the imports

File: nebullvm/optimizers/tensor_rt.py [hunk 4]
```diff
@@ -71,7 +93,16 @@ def _build_and_save_the_engine(
...
+        try:
+            config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, 1 << 30)
+        except AttributeError:
+            # The method set_memory_pool_limit is not available
+            # until TensorRT Release 8.4.1
+            warnings.warn(
+                "Cannot call method set_memory_pool_limit for TensorRT."
+                "Please update TensorRT version."
+            )
+            pass
```
Comment (morgoth95) [inline nebullvm/optimizers/tensor_rt.py:105]: do we need this extra `pass` after the warning?

File: nebullvm/optimizers/tensor_rt.py [hunk 5]
```diff
@@ -246,3 +277,85 @@ def optimize(
...
+            f"Optimizing with {self.__class__.__name__} and "
+            f"q_type: {quantization_type}."
+        )
+        if not torch.cuda.is_available():
+            raise SystemError(
+                "You are trying to run an optimizer developed for NVidia gpus "
+                "on a machine not connected to any GPU supporting CUDA."
+            )
+
+        supported_q_types = [
```
Comment (morgoth95) [inline nebullvm/optimizers/tensor_rt.py:301]: Does TensorRT's PyTorch interface accept only full precision and half precision? ONNX's also supports 8-bit quantization! I would be quite surprised if this is not the case for PyTorch as well.

File: nebullvm/optimizers/tests/utils.py [hunk 6]
```diff
@@ -3,8 +3,12 @@
 
 import torch
 
-from nebullvm.base import ModelParams
+from nebullvm.api.functions import _extract_info_from_data
+from nebullvm.api.huggingface import convert_hf_model
+from nebullvm.base import ModelParams, DeepLearningFramework
 from nebullvm.converters.torch_converters import convert_torch_to_onnx
+from nebullvm.utils.data import DataManager
+from transformers import AlbertModel, AlbertTokenizer
```
Comment (morgoth95) [inline nebullvm/optimizers/tests/utils.py:11]: sort the imports

**Review by morgoth95 — CHANGES_REQUESTED** [review]

File: nebullvm/inference_learners/tensor_rt.py (hunk 2 above)
Comment (valeriosofi) [reply nebullvm/inference_learners/tensor_rt.py:254]: Yep, the model returned after compiling with torch_tensorrt is in TorchScript format

File: nebullvm/inference_learners/tensor_rt.py [hunk 7]
```diff
@@ -248,6 +247,59 @@ def load(cls, path: Union[Path, str], **kwargs):
...
+        self.model = torch_model.eval()
+        if torch.cuda.is_available():
+            self.model.cuda()
+        self.dtype = dtype
+
+    def run(self, *input_tensors: torch.Tensor) -> Tuple[torch.Tensor, ...]:
+        device = input_tensors[0].device
+        if torch.cuda.is_available():
+            if self.dtype == torch.half:
+                input_tensors = (t.cuda().half() for t in input_tensors)
```
Comment (morgoth95) [inline nebullvm/inference_learners/tensor_rt.py:266]: Note that model.half() does not change the embedding parameters (for instance in the first transformer layer), so casting the input to half can give an error when using half precision with transformers. Just float32 inputs should be converted to half precision.

Comment (morgoth95) [reply nebullvm/inference_learners/tensor_rt.py:266]: Note that this should also be reflected in the HalfPrecisionTransformation (in nebullvm.transformations.precision_tfms).

File: nebullvm/optimizers/tensor_rt.py [hunk 8]
```diff
@@ -47,6 +51,25 @@
                 "No TensorRT valid installation has been found. "
                 "It won't be possible to use it in the following."
             )
+    try:
+        import torch_tensorrt
+    except ImportError:
+        # TODO: Remove the False flag for allowing Torch-TensorRT to
+        #  be installed by the Auto-Installer.
+        if False and not NO_COMPILER_INSTALLATION:
```
Comment (morgoth95) [inline nebullvm/optimizers/tensor_rt.py:59]: I think we can activate this installation by default. Just check if cuda is available.

File: nebullvm/optimizers/tests/test_huggingface.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+from tempfile import TemporaryDirectory
+
+import pytest
+
+from nebullvm.base import DeepLearningFramework
+from nebullvm.api.huggingface import HuggingFaceInferenceLearner
```
Comment (morgoth95) [inline nebullvm/optimizers/tests/test_huggingface.py:6]: sort the imports

File: nebullvm/optimizers/tests/utils.py (hunk 6 above)
Comment (morgoth95) [inline nebullvm/optimizers/tests/utils.py:11]: `transformers` should be imported on the block above (after PyTorch)

File: nebullvm/utils/data.py [hunk 10]
```diff
@@ -99,3 +100,26 @@ def split(self, split_pct: float, shuffle: bool = False):
             DataManager([self[i] for i in idx[:n]]),
             DataManager([self[i] for i in idx[n:]]),
         )
+
+
+class PytorchDataset(torch.utils.data.Dataset):
+    def __init__(self, input_data: DataManager):
+        self.data = input_data
+        self.batch_size = input_data[0][0][0].shape[0]
+        return
```
Comment (morgoth95) [inline nebullvm/utils/data.py:109]: not needed

File: nebullvm/utils/data.py [hunk 11]
```diff
@@ -99,3 +100,26 @@ def split(self, split_pct: float, shuffle: bool = False):
...
+
+    def __len__(self):
+        return sum(
+            [
+                d[0].shape[0]
+                for d in [
+                    [inputs for inputs in batch[0]] for batch in self.data
+                ]
+            ]
+        )
```
Comment (morgoth95) [inline nebullvm/utils/data.py:119]:
```suggestion
        return sum(
            [
                batch_inputs[0].shape[0]
                for batch_inputs, _ in self.data
            ]  # we assume the batch size being the first dim of the first input tensor
        )
```

File: nebullvm/utils/torch.py [hunk 12]
```diff
@@ -37,11 +37,16 @@ def create_model_inputs_torch(
...
+    input_tensors: List[torch.Tensor],
+    dtype: torch.dtype = torch.float32,
 ) -> List[torch.Tensor]:
     if torch.cuda.is_available():
         torch_model.cuda()
-        input_tensors = (t.cuda() for t in input_tensors)
+        if dtype != torch.half:
+            input_tensors = (t.cuda() for t in input_tensors)
+        else:
+            input_tensors = (t.cuda().half() for t in input_tensors)
```
Comment (morgoth95) [inline nebullvm/utils/torch.py:49]: same as before

**Review by morgoth95 — CHANGES_REQUESTED** [review]

File: nebullvm/utils/data.py (hunk 11 above)
Comment (valeriosofi) [reply nebullvm/utils/data.py:119]: Yep I agree, it's definitely better in this way

File: nebullvm/inference_learners/tensor_rt.py [hunk 13]
```diff
@@ -248,6 +247,63 @@ def load(cls, path: Union[Path, str], **kwargs):
...
+        self.dtype = dtype
+
+    def run(self, *input_tensors: torch.Tensor) -> Tuple[torch.Tensor, ...]:
+        device = input_tensors[0].device
+        if torch.cuda.is_available():
+            if self.dtype == torch.half:
+                input_tensors = (
+                    t.cuda().half()
+                    for t in input_tensors
+                    if t.dtype == torch.float32
```
Comment (morgoth95) [inline nebullvm/inference_learners/tensor_rt.py:269]: what if it is int32? ;)

**Pull request merged**